qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+from abc import abstractmethod, ABC
+from dataclasses import dataclass, field
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
+
+from .optimizer import Optimizer, POINT, OptimizerResult
+import numpy as np
+
+CALLBACK = Callable[[int, np.ndarray, float, float], None]
+
+
+@dataclass
+class AskObject(ABC):
+    """Base class for return type of method SteppableOptimizer.ask()"""
+
+    x_fun: Union[POINT, List[POINT]]
+    x_jac: Union[POINT, List[POINT]]
+
+
+@dataclass
+class TellObject(ABC):
+    """Base class for argument type of method SteppableOptimizer.tell()"""
+
+    eval_fun: Union[float, List[float]]
+    eval_jac: Union[POINT, List[POINT]]
+
+
+@dataclass
+class OptimizerState:
+    """
+    Base class representing the state of the optmiizer.
+    Any variable that changes during the optimization process and is needed for the next step
+    of optimization should be stored in this dataclass.
+    """
+
+    x: POINT  # pylint: disable=invalid-name
+    fun: Optional[Callable[[POINT], float]]  # Make optional
+    jac: Optional[Callable[[POINT], POINT]]
+    nfev: Optional[int]
+    njev: Optional[int]
+    nit: Optional[int]
+
+
+class SteppableOptimizer(Optimizer):
+    """
+    Base class for a steppable optimizer.
+
+    This family of optimizers will be using the ask and tell interface.
```

Comment:
Could you reference the link from below already here? 
```suggestion
    This family of optimizers will be using the 
    `ask and tell interface <https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html>`_.
```